### PR TITLE
[webview_flutter] Add a method for getting the user agent

### PR DESCRIPTION
**Pull request by bparrishMines**
Aggregate PR for adding a method for retrieving the user agent. 

Internal: b/272582342

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/webview_flutter/webview_flutter_android/test/android_webview_controller_test.dart [hunk 1]
```diff
@@ -1056,6 +1056,17 @@ void main() {
...
       verify(mockSettings.setUserAgentString('Test Framework')).called(1);
     });
+
+    test('getUserAgent', () async {
+      final MockWebSettings mockSettings = MockWebSettings();
+      final AndroidWebViewController controller = createControllerWithMocks(
+        mockSettings: mockSettings,
+      );
+      when(mockSettings.getUserAgentString())
+          .thenAnswer((_) => Future<String?>.value('str'));
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/test/android_webview_controller_test.dart:1066]: Nit: It would be clearer to use a `const String` local variable for this and the expectation, to make it obvious that what's expected is that the mock value is being returned exactly.

File: packages/webview_flutter/webview_flutter_wkwebview/test/src/web_kit/web_kit_test.dart [hunk 2]
```diff
@@ -833,14 +833,21 @@ void main() {
...
           webViewInstanceId,
           'hello',
         ));
       });
 
+      test('getCustomUserAgent', () {
+        when(
+          mockPlatformHostApi.getCustomUserAgent(webViewInstanceId),
+        ).thenReturn('str');
+        expect(webView.getCustomUserAgent(), completion('str'));
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/test/src/web_kit/web_kit_test.dart:848]: See comment in the Android test about using a local constant.

File: packages/webview_flutter/webview_flutter_wkwebview/test/webkit_webview_controller_test.dart [hunk 3]
```diff
@@ -823,6 +823,19 @@ void main() {
...
+      final MockWKWebView mockWebView = MockWKWebView();
+
+      final WebKitWebViewController controller = createControllerWithMocks(
+        createMockWebView: (_, {dynamic observeValue}) => mockWebView,
+      );
+
+      when(mockWebView.getCustomUserAgent()).thenAnswer(
+        (_) => Future<String?>.value('str'),
+      );
+      expect(controller.getUserAgent(), completion('str'));
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/test/webkit_webview_controller_test.dart:836]: See comment in the Android test about using a local constant.

Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/test/webkit_webview_controller_test.dart:836]: See comment in the Android test about using a local constant.

File: packages/webview_flutter/webview_flutter/test/webview_controller_test.dart [hunk 4]
```diff
@@ -387,6 +387,20 @@ void main() {
...
+        MockPlatformWebViewController();
+    when(mockPlatformWebViewController.getUserAgent()).thenAnswer(
+      (_) => Future<String?>.value('str'),
+    );
+
+    final WebViewController webViewController = WebViewController.fromPlatform(
+      mockPlatformWebViewController,
+    );
+
+    await expectLater(webViewController.getUserAgent(), completion('str'));
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter/test/webview_controller_test.dart:402]: See comment in the Android test about using a local constant.

File: packages/webview_flutter/webview_flutter_android/test/android_webview_test.dart [hunk 5]
```diff
@@ -527,6 +527,12 @@ void main() {
           100,
         ));
       });
+
+      test('getUserAgentString', () async {
+        when(mockPlatformHostApi.getUserAgentString(webSettingsInstanceId))
+            .thenReturn('str');
+        expect(await webSettings.getUserAgentString(), 'str');
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/test/android_webview_test.dart:534]: See comment in the other Android test about using a local constant.

File: packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebSettingsHostApiImpl.java [hunk 6]
```diff
@@ -132,4 +132,11 @@ public void setTextZoom(@NonNull Long instanceId, @NonNull Long textZoom) {
     final WebSettings webSettings = Objects.requireNonNull(instanceManager.getInstance(instanceId));
     webSettings.setTextZoom(textZoom.intValue());
   }
+
+  @Nullable
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebSettingsHostApiImpl.java:136]: Does the underlying API really return a nullable string?

File: packages/webview_flutter/webview_flutter_android/android/src/test/java/io/flutter/plugins/webviewflutter/WebSettingsTest.java [hunk 7]
```diff
@@ -117,4 +118,10 @@ public void setTextZoom() {
     testHostApiImpl.setTextZoom(0L, 100L);
     verify(mockWebSettings).setTextZoom(100);
   }
+
+  @Test
+  public void getWebSettings() {
+    when(mockWebSettings.getUserAgentString()).thenReturn("str");
+    assertEquals(testHostApiImpl.getUserAgentString(0L), "str");
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/android/src/test/java/io/flutter/plugins/webviewflutter/WebSettingsTest.java:125]: See comment in the other Android test about using a local constant.

File: packages/webview_flutter/webview_flutter_wkwebview/example/ios/RunnerTests/FWFWebViewHostApiTests.m [hunk 8]
```diff
@@ -480,4 +480,20 @@ - (void)testSetInspectable API_AVAILABLE(ios(16.4), macos(13.3)) {
...
+
+  FWFInstanceManager *instanceManager = [[FWFInstanceManager alloc] init];
+  [instanceManager addDartCreatedInstance:mockWebView withIdentifier:0];
+
+  FWFWebViewHostApiImpl *hostAPI = [[FWFWebViewHostApiImpl alloc]
+      initWithBinaryMessenger:OCMProtocolMock(@protocol(FlutterBinaryMessenger))
+              instanceManager:instanceManager];
+
+  FlutterError *error;
+  XCTAssertEqualObjects([hostAPI userAgentForWebViewWithIdentifier:@0 error:&error], @"str");
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/example/ios/RunnerTests/FWFWebViewHostApiTests.m:496]: See comment in the Android test about using a local constant.

File: packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart [hunk 9]
```diff
@@ -697,6 +697,9 @@ abstract class WKWebViewHostApi {
 
   @ObjCSelector('setInspectableForWebViewWithIdentifier:inspectable:')
   void setInspectable(int identifier, bool inspectable);
+
+  @ObjCSelector('userAgentForWebViewWithIdentifier:')
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart:701]: This should be `customUserAgentForWebViewWithIdentifier`, since it doesn't give you the actual user agent.

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart [hunk 10]
```diff
@@ -558,6 +558,9 @@ class WebKitWebViewController extends PlatformWebViewController {
   Future<void> setInspectable(bool inspectable) {
     return _webView.setInspectable(inspectable);
   }
+
+  @override
+  Future<String?> getUserAgent() => _webView.getCustomUserAgent();
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart:563]: These aren't equivalent; the RHS returns null unless a custom user agent has been set, so this implementation would give people the wrong answer on iOS (except in cases where they set it themselves, but they probably don't need to ask then). It looks like we need to fall back to using JS when this returns null.

Comment (stuartmorgan) [reply packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart:563]: This issue also suggests we need an integration test that validates that we get a non-null value when we haven't set a custom agent.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by ditman** [comment]
I'll keep an eye out for when this lands, the `_web` version of this should be very simple to add!

File: packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebSettingsHostApiImpl.java (hunk 6 above)
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebSettingsHostApiImpl.java]: I tested this out and it looks like it is always nonnull. Updated.

**Review by stuartmorgan — APPROVED** [review]
LGTM

File: packages/webview_flutter/webview_flutter_android/test/legacy/webview_android_widget_test.mocks.dart [hunk 11]
```diff
@@ -303,6 +303,14 @@ class MockWebSettings extends _i1.Mock implements _i2.WebSettings {
...
 
   @override
+  _i5.Future<String> getUserAgentString() => (super.noSuchMethod(
+        Invocation.method(
+          #getUserAgentString,
+          [],
+        ),
+        returnValue: _i5.Future<String>.value(''),
+      ) as _i5.Future<String>);
+  @override
```
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/test/legacy/webview_android_widget_test.mocks.dart:313]:
```suggestion
```
